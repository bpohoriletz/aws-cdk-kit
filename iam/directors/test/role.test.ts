import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { arn } from '../../../test/stubs';
import RoleDirector from '../role';
import EbEc2RoleBuilder from '../../role-builders/eb-ec2-role-builder';
import CodedeployRoleBuilder from '../../role-builders/codedeploy-role-builder';
import GithubActionsRoleBuilder from '../../role-builders/github-actions-role-builder';

describe('new RoleDirector()', () => {
  let instance: RoleDirector;
  let stack: cdk.Stack;

  beforeEach(() => {
    stack = new cdk.Stack();
  });

  it('constructs role for ElasticBeanstalk EC2 instances', () => {
    instance = new RoleDirector(EbEc2RoleBuilder);
    instance.constructEc2Role(stack, 'Ec2Role', arn());

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });

  it('constructs role for CodeDeploy', () => {
    instance = new RoleDirector(CodedeployRoleBuilder);
    instance.constructCodedeployRole(stack, 'CodeDeployRole', arn());

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });

  it('constructs role for GitHub Actions', () => {
    instance = new RoleDirector(GithubActionsRoleBuilder);
    instance.constructGtihubRole(stack, 'github_acc');

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
