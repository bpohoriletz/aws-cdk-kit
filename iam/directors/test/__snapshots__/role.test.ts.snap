// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new RoleDirector() constructs role for CodeDeploy 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CodeDeployRole12BEECBE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codedeploy.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSCodeDeployRole",
              ],
            ],
          },
        ],
        "RoleName": "CodeDeployRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "allowEc2UploadLogsToCodedeployRegionalBucketF9632D95": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:ListBucket",
                "s3:ListBucketVersions",
                "s3:GetObject",
                "s3:GetObjectVersion",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:partition:service:region:account-id:resource",
                "arn:partition:service:region:account-id:resource/resources/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2UploadLogsToCodedeployRegionalBucketF9632D95",
        "Roles": [
          {
            "Ref": "CodeDeployRole12BEECBE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`new RoleDirector() constructs role for ElasticBeanstalk EC2 instances 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Ec2Role2FD9A272": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "allowEc2PutStatisticsD9C1A435": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "elasticbeanstalk:PutInstanceStatistics",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2PutStatisticsD9C1A435",
        "Roles": [
          {
            "Ref": "Ec2Role2FD9A272",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "allowEc2UploadLogsA49BD994": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2UploadLogsA49BD994",
        "Roles": [
          {
            "Ref": "Ec2Role2FD9A272",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "allowEc2UploadLogsToEbRegionalBucket40230F17": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:ListBucket",
                "s3:ListBucketVersions",
                "s3:GetObject",
                "s3:GetObjectVersion",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:partition:service:region:account-id:resource",
                "arn:partition:service:region:account-id:resource/resources/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2UploadLogsToEbRegionalBucket40230F17",
        "Roles": [
          {
            "Ref": "Ec2Role2FD9A272",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`new RoleDirector() constructs role for GitHub Actions 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GithubActionsRoleF5CC769F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": "repo:github_acc/*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":oidc-provider/token.actions.githubusercontent.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role for GitHub Actions to access AWS resources",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSElasticBeanstalkWebTier",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/SecretsManagerReadWrite",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSElasticBeanstalkManagedUpdatesCustomerRolePolicy",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "accessCodeDeployBucket0952D2BF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:ListBucket",
                "s3:ListBucketVersions",
                "s3:GetObject",
                "s3:GetObjectVersion",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:aws:s3:::codedeploy-*",
                "arn:aws:s3:::codedeploy-*/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "accessCodeDeployBucket0952D2BF",
        "Roles": [
          {
            "Ref": "GithubActionsRoleF5CC769F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "codeDeploy97BF7F9E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codedeploy:*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "codeDeploy97BF7F9E",
        "Roles": [
          {
            "Ref": "GithubActionsRoleF5CC769F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
